src/pandablocks/asyncio.py

```diff
@@ -132,48 +132,61 @@ async def data(
         scaled: bool = True,
         flush_period: Optional[float] = None,
         frame_timeout: Optional[float] = None,
+        flush_event: Optional[asyncio.Event] = None,
     ) -> AsyncGenerator[Data, None]:
         """Connect to data port and yield data frames
 
         Args:
             scaled: Whether to scale and average data frames, reduces throughput
             flush_period: How often to flush partial data frames, None is on every
-                chunk of data from the server
+                chunk of data from the server. 0 is never and flushing will be
```

Comment:
I'm not a big fan of using both `None` and `0` to be different things, although I accept that this is likely a stylistic preference. Is this particular interface desirable due to how it will be used elsewhere?

Is it cleaner to have a separate `flush_type` parameter, with Enum values of `Automatic` and `Manual`? `flush_period` is then not being overloaded to mean both how often data is flushed and whether it's automatic or not

Reply:
With the current PVs, we need `0` to be no periodic flush, as the user needs to be able to put that in through the epics field `:FlushPeriod` (which won't accept `None`).

I think the alternative would be to have a new boolOut pv called `:PeriodicFlushOn`. Setting that `0` will turn off periodic flush and `:FlushPeriod` of 0 will cause flushing on every frame. The more I think about it the more I like this... what do you think?

Reply:
In this branch currently `0` is no periodic flush only so that the user can turn off automatic flushing with the `:FlushPeriod` box. I prefer the flush_type idea. I'll implement that and have `0` flush period be flushing every frame.

Reply:
I'm missing something: how would EPICS have ever provided the `None` value? 

I'm not convinced its a good idea to write Python code that could be consumed by anyone and specifically tailor it to EPICS conventions/restrictions. I suggest that the cleanest, most understandable Python code would be to add a new `flush_type` parameter, and we deal with how the PVs that provide that information look as a separate discussion.

Reply:
This was my suggestion to avoid breaking back-compatibility. At the moment we have 2 options:
- `flush_period is None`: flush every packet from PandA
- `flush_period > 0`: flush every `flush_period` seconds

We want to add a third option to flush on an Event, so I suggested using `flush_period==0` to control the periodic flush. Thinking about this, we can get away without having `0` and `None` be different things with the following:
- If `flush_period` then flush every `flush_seconds`
- If `flush_event` then flush whenever requested
- If neither is supplied then flush every frame

This keeps back-compat, but lets us do manual only flushing by supplying only `flush_event` and not `flush_period`

What do you reckon?

Reply:
That solution still has the unfortunate issue that if you specify both a non-zero `flush_period` and also provide a `flush_event`, you are doing manual flushing whether you intended to or not. We could of course make providing both an error, which would solve that one. But I still feel it's somewhat indicative that we're making a mistake trying to overload this one function to do both things.

I take your point that it would be a breaking API change to do what I'm suggesting.

Reply:
It was my intention while writing this that the manual flush would work even if `flush_period > 0`. A user should be able to flush manually even if the system is automatically flushing every n seconds. Do you think the use-case for this is too limited to justify this slightly more complicated behavior?

Reply:
I'll have to leave that question to @coretl

Reply:
I'll talk to him in person, I would argue for the following:

* if `flush_periodically` (new PV tied to `boolOut`) then flush every `flush_seconds`/`flush_period` (tied to `analogueOut`).
    * also flush whenever requested by `flush_event`. If `flush_seconds`/`flush_period` is 0 then this is the only way to flush.
* If not `flush_periodically` then flush every frame, `flush_event` has no effect.

Reply:
Ok, how about this:
- `flush_mode: Enum("Immediate", "Periodic", "Manual") = "Immediate"`
- `flush_period: Optional[float] = None`: In `Periodic` mode this must be >0 and starts a periodic task, otherwise it must be None
- `flush_event: Optional[Event] = None`: In `Manual` mode this must supply an `Event` which a periodic task waits on for flush, otherwise it must be None

Epics interface is 3 PVs, `FlushMode` mbbo, `FlushPeriod` ao, `FlushNow` action.

Reply:
This keeps back compat for the default case, and will error with a suitable message if a `flush_period` is supplied but mode is not changed

Reply:
I think that sounds good me to!